[PTRun] Drag and drop files

daniel-richter opened the pull request:
## Summary of the Pull Request
Support drag&drop to other application for files in result list. 
![Recording 2022-12-01 at 20 03 27](https://user-images.githubusercontent.com/12428377/205138315-71d8f802-de20-4c6f-91d1-3b2c5e556b44.gif)

## PR Checklist

- [x] **Closes:** #4462
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Dev docs:** Added/updated

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
To implement this behaviour, I added the following things:
- detect dragging an item from the suggestions list
- check whether this item represents a file/folder (something path-related)
- perform a DragDrop with path as FileDrop
- make related plugins (that return files/folders) mark their results as draggable file/folder

### detect dragging an item from the suggestions list
extend _src/modules/launcher/PowerLauncher/MainWindow.xaml.cs_
Save mouse position on mouse down and check aggainst system drag distance
```cs
private void SuggestionsList_PreviewMouseLeftButtonDown(object sender, MouseButtonEventArgs e)
{
    _mouseDownPosition = e.GetPosition(null);
    _mouseDownResultViewModel = ((FrameworkElement)e.OriginalSource).DataContext as ResultViewModel;
}

private void SuggestionsList_MouseMove(object sender, MouseEventArgs e)
{
    if (e.LeftButton == MouseButtonState.Pressed && ...)
    {
        Vector dragDistance = _mouseDownPosition - e.GetPosition(null);
        if (Math.Abs(dragDistance.X) > SystemParameters.MinimumHorizontalDragDistance || Math.Abs(dragDistance.Y) > SystemParameters.MinimumVerticalDragDistance)
        {
            ...
       }
    }
}
```

### check whether this item represents a file/folder (something path-related)
Since all plugins that return files/folders put their data objects containing the path information into `Result.ContextData`, I decided to introduce an interface `IFileDropResult` to indicate that a result contains a file/folder to be draggable.
_src/modules/launcher/Wox.Plugin/Interfaces/IFileDropResult.cs_
```cs
public interface IFileDropResult
{
    public string Path { get; set; }
}
```
So we can check whether the result connected to the view model of the clicked result item implements this interface.

```cs
private void SuggestionsList_MouseMove(object sender, MouseEventArgs e)
{
    if (... && _mouseDownResultViewModel?.Result?.ContextData is IFileDropResult fileDropResult)
    {  ...  }
}
```

### perform a DragDrop with path as FileDrop

If so, the PTRun window will be hidden and the corresponding path of the dragged item will be made available for dragging as FileDrop:
```cs
_viewModel.Hide();
DataObject dataObject = new DataObject(DataFormats.FileDrop, new[] { fileDropResult.Path });
DragDrop.DoDragDrop(ListBox.SuggestionsList, dataObject, DragDropEffects.Copy);
```

### make related plugins (that return files/folders) mark their results as draggable file/folder

The search result types of the following plugins now implement `IFileDropResult`:
- **Folder**/SearchResult.cs
- **Indexer**/SearchHelper/SearchResult.cs
- ([**Everything**/SearchResult.cs](https://github.com/lin-ycv/EverythingPowerToys/blob/main/SearchHelper/SearchResult.cs))

For unification, I had to rename Folder/SearchResult's `FullPath` to `Path`.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

(See screen recording)

![Recording 2022-12-01 at 19 59 50](https://user-images.githubusercontent.com/12428377/205138356-6cafc2e6-1083-4f25-a209-b1372b081bc4.gif)

daniel-richter committed:
[PTRun] Support drag&drop to other application for files in result list

daniel-richter commented:
@microsoft-github-policy-service agree

htcfreek commented:
Is this feature only for folder plugin? It might make sense to have it in the indexer plugin too.

daniel-richter commented:
As already stated above, this feature is available for the following plugins:

> - **Folder**/SearchResult.cs
> - **Indexer**/SearchHelper/SearchResult.cs
> - ([**Everything**/SearchResult.cs](https://github.com/lin-ycv/EverythingPowerToys/blob/main/SearchHelper/SearchResult.cs))

htcfreek commented:
> As already stated above, this feature is available for the following plugins:
> 
> > - **Folder**/SearchResult.cs
> > - **Indexer**/SearchHelper/SearchResult.cs
> > - ([**Everything**/SearchResult.cs](https://github.com/lin-ycv/EverythingPowerToys/blob/main/SearchHelper/SearchResult.cs))

But is it correct that the everything plugin needs to be updated to implement the new interface?

daniel-richter commented:
Yes. But I can't do it without having this changes in PTRun first. :-)

htcfreek commented:
Do we have a check if the access to the path is denied. We should prevent a crash in this case.

daniel-richter commented:
> Do we have a check if the access to the path is denied. We should prevent a crash in this case.

Hm, not sure what to do... The Drag&Drop does not process "the file" in any way but only passes the path (string) as data to the shell. So `DoDragDrop` should not fail, because what the target does with the path is the responsibility of the drop target. The transfer of the data (path as string) itself is successful if a target accepts the drop. A drop target application does not even have to open access the path but only process the provided path in some way (e.g. add to text).

**Open questions:**

1. Should return value of `DragDrop.DoDragDrop(ListBox.SuggestionsList, dataObject, DragDropEffects.Copy);` be checked in any way? `DoDragDrop` will block the execution until drag&drop is finished - the result value is `DragDropEffects.Copy` if some drop target accepted the drop or `DragDropEffects.None` if there wasn't.
Maybe make the search results visible aggain if there was no drop?

2. An alternative implementation to indicate file/folder search results (current proposal: a special type (`IFileDropResult`) for the `Result.ContextData` property) is to make the file/folder path a first class property of `Result`. Would that be more elgant?

3. I have a working solution for adding the file thumbnail as drop image. Should I include that into this pull request?

htcfreek commented:
> > Do we have a check if the access to the path is denied. We should prevent a crash in this case.
> 
> Hm, not sure what to do... The Drag&Drop does not process "the file" in any way but only passes the path (string) as data to the shell. So `DoDragDrop` should not fail, because what the target does with the path is the responsibility of the drop target. The transfer of the data (path as string) itself is successful if a target accepts the drop. A drop target application does not even have to open access the path but only process the provided path in some way (e.g. add to text).
> 
> **Open questions:**
> 
> 1. Should return value of `DragDrop.DoDragDrop(ListBox.SuggestionsList, dataObject, DragDropEffects.Copy);` be checked in any way? `DoDragDrop` will block the execution until drag&drop is finished - the result value is `DragDropEffects.Copy` if some drop target accepted the drop or `DragDropEffects.None` if there wasn't.
> Maybe make the search results visible aggain if there was no drop?
> 
> 2. An alternative implementation to indicate file/folder search results (current proposal: a special type (`IFileDropResult`) for the `Result.ContextData` property) is to make the file/folder path a first class property of `Result`. Would that be more elgant?
> 
> 3. I have a working solution for adding the file thumbnail as drop image. Should I include that into this pull request?
> 
> 

Regarding the error check: If we can't crash or hang I see no reason to handle something in our code.

Regarding the thumbnail: Do you have a screenshot of what you mean?

daniel-richter commented:
| no drag image | with file thumbnail as drag image |
| --- | --- |
| ![Recording 2022-12-02 at 16 16 17](https://user-images.githubusercontent.com/12428377/205325962-9ea61c2e-5b6f-4301-a1d2-df3fe7fdc673.gif) | ![Recording 2022-12-02 at 16 18 16](https://user-images.githubusercontent.com/12428377/205326019-6694303b-8973-4839-af5d-cf94bc730e6f.gif) |

htcfreek commented:
> | no drag image | with file thumbnail as drag image |
> | --- | --- |
> | ![Recording 2022-12-02 at 16 16 17](https://user-images.githubusercontent.com/12428377/205325962-9ea61c2e-5b6f-4301-a1d2-df3fe7fdc673.gif) | ![Recording 2022-12-02 at 16 18 16](https://user-images.githubusercontent.com/12428377/205326019-6694303b-8973-4839-af5d-cf94bc730e6f.gif) |
> 
> Note: Displayingd the drag image is also the responsibility of the drop target. In the shown example it's Windows Explorer.
> 
> 
> 

Looks good. Let's push it.

daniel-richter committed:
[PTRun] use file/folder thumbnail as drag image

daniel-richter commented:
> Looks good. Let's push it.

Done. Changed `DoDragDrop` to:

```cs
_viewModel.Hide();

try
{
    // DoDragDrop with file thumbnail as drag image
    var dataObject = DragDataObject.FromFile(fileDropResult.Path);
    IntPtr hBitmap = Image.WindowsThumbnailProvider.GetHBitmap(Path.GetFullPath(fileDropResult.Path), Constant.ThumbnailSize, Constant.ThumbnailSize, Image.ThumbnailOptions.None);
    try
    {
        dataObject.SetDragImage(hBitmap, Constant.ThumbnailSize, Constant.ThumbnailSize);
        DragDrop.DoDragDrop(ListBox.SuggestionsList, dataObject, DragDropEffects.Copy);
    }
    finally
    {
        Image.NativeMethods.DeleteObject(hBitmap);
    }
}
catch
{
    // DoDragDrop without drag image
    IDataObject dataObject = new DataObject(DataFormats.FileDrop, new[] { fileDropResult.Path });
    DragDrop.DoDragDrop(ListBox.SuggestionsList, dataObject, DragDropEffects.Copy);
}
```
I added _src/modules/launcher/PowerLauncher/Helper/DragDataObject.cs_ with helper methods to deal with shell data objects (and drag images) and had to make `Image.WindowsThumbnailProvider.GetHBitmap` public (because a HBitmap/IntPtr is needed for the thumbnail image).

daniel-richter committed:
(fix spellcheck)

daniel-richter commented:
> When do we create the hBitmap? Can this make search or dragging slow?

It's done before starting the drag&drop action (after clicking one result item and dragging the mouse the minumum drag distance). So drag&drop function won't affect searching in any way. 

Creating the hBitmap could be slow, but the _Folder_ and _Indexer_ plugins will set the `IconPath` to the file/folder, so the thumbnails will be loaded anyway. (The _Everything_ plugin has a setting do diable preview.)

The thumbnails for search results will be cached (`ImageSource Wox.Infrastructure.Image.ImageLoader.Load`). Also, the drag icon itself is already present in `PowerLauncher.ViewModel.ResultViewModel.Image` resp. `_mouseDownResultViewModel.Image`. But I couldn't find a nice way get an hBitmap from a`n `ImageSource` typed object.
If that could be done, we could simply reuse the already loaded icon from the search result.

daniel-richter committed:
[PTRun] use _mouseDownResultViewModel.Image to generate the drag image

daniel-richter commented:
> [...] the drag icon itself is already present in `PowerLauncher.ViewModel.ResultViewModel.Image` resp. `_mouseDownResultViewModel.Image`. But I couldn't find a nice way get an hBitmap from a`n `ImageSource` typed object. If that could be done, we could simply reuse the already loaded icon from the search result.

Okay, found a solution in https://weblog.west-wind.com/posts/2020/Sep/16/Retrieving-Images-from-the-Clipboard-and-WPF-Image-Control-Woes
Now, `PowerLauncher.ViewModel.ResultViewModel.Image` is (re)used for the drag image.

daniel-richter committed:
fix spelling + refactoring

jaimecbernardo reviewed (approved):
Pretty nice feature, and it's working very well!
Thanks a lot for the contribution!
